src/Raven.Server/Web/System/AdminDatabasesHandler.cs

```diff
@@ -352,6 +357,69 @@ public Task GetPeriodicBackupStatus()
             return Task.CompletedTask;
         }
 
+        [RavenAction("/admin/database-restore", "POST", "/admin/database-restore")]
+        public async Task RestoreDatabase()
+        {
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            //using (ServerStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+            {
+                context.OpenReadTransaction();
+
+                var operationId = GetLongQueryString("operationId", required: false);
+
+                var restoreConfiguration = await context.ReadForMemoryAsync(RequestBodyStream(), "database-restore");
+                var restoreConfigurationJson = JsonDeserializationCluster.RestoreBackupConfiguraionConfiguration(restoreConfiguration);
+
+                var databaseName = restoreConfigurationJson.DatabaseName;
+                if (string.IsNullOrWhiteSpace(databaseName))
+                    throw new ArgumentException("Database name can't be null or empty");
+
+                string errorMessage;
+                if (ResourceNameValidator.IsValidResourceName(databaseName, ServerStore.Configuration.Core.DataDirectory.FullPath, out errorMessage) == false)
+                    throw new BadRequestException(errorMessage);
+
+                if (ServerStore.DatabasesLandlord.DatabasesCache.TryGetValue(new StringSegment(databaseName), out _))
+                    throw new ArgumentException($"Cannot restore data to an existing database named {databaseName}");
+
+                // TODO: prevent the creation of a database with the same name during restore
+
+                using (var token = new OperationCancelToken(ServerStore.ServerShutdown))
+                {
+                    var restoreBackupTask = new RestoreBackupTask(
+                        ServerStore, 
+                        restoreConfigurationJson, 
+                        context, 
+                        isEncrypted => AssignNodesToDatabase(context, 1, restoreConfigurationJson.DatabaseName, isEncrypted, out _), 
+                        token.Token);
+
+                    if (operationId.HasValue)
+                    {
+                        try
+                        {
+                            await
+                                ServerStore.Operations.AddOperation(
+                                    "Restoring database: " + databaseName,
+                                    Documents.Operations.Operations.OperationType.DatabaseRestore,
+                                    taskFactory: onProgress => Task.Run(async() => await restoreBackupTask.Execute(onProgress), token.Token), id: operationId.Value, token: token);
+                        }
+                        catch (Exception)
+                        {
+                            HttpContext.Abort();
```

Comment:
Why abort? Shouldn't we be sending the error to the user? What happens otherwise, how does they know about an issue?